File: weisi/src/main/java/com/cjq/lib/weisi/communicator/tcp/TcpServer.java
package com.cjq.lib.weisi.communicator.tcp;

import android.support.annotation.NonNull;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;

/**
 * Created by CJQ on 2018/3/9.
 */

public class TcpServer extends Tcp {

    private ServerSocket mSocket;

    public boolean launch(int localPort) {
        if (!isLaunched()) {
            try {
                mSocket = new ServerSocket(localPort);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return isLaunched();
    }

    public boolean isLaunched() {
        return mSocket != null;
    }

    public void accept(@NonNull final OnClientAcceptListener listener) {
        if (listener != null && isLaunched() && mState != CONNECTING) {
            mState = CONNECTING;
            new Thread(new Runnable() {
                @Override
                public void run() {
                    while (isLaunched() && mState == CONNECTING) {
                        TcpSocket socket = null;
                        try {
                            Socket client = mSocket.accept();
                            socket = new TcpSocket(client);
                        } catch (IOException e) {
                            e.printStackTrace();
                        } finally {
                            mState = socket != null ? CONNECTED : UNCONNECTED;
                            if (listener.onClientAccept(mState, socket)) {
                                mState = CONNECTING;
                            }
                        }
                    }
                }
            }).start();
        } else {
            listener.onClientAccept(mState, null);
        }
    }

    public void shutdown() {
        if (isLaunched()) {
            try {
                mSocket.close();
            } catch (IOException e) {
                e.printStackTrace();
            } finally {
                mState = UNCONNECTED;
            }
        }
    }

    /**
     * Created by CJQ on 2018/3/12.
     */
    public interface OnClientAcceptListener {
        //返回true将继续等待客户端连接
        boolean onClientAccept(@ConnectState int state, TcpSocket socket);
    }
}

File: wsb/src/main/java/com/wsn/lib/wsb/communicator/tcp/TcpServer.kt
package com.wsn.lib.wsb.communicator.tcp

import java.io.IOException
import java.net.ServerSocket

import com.wsn.lib.wsb.communicator.tcp.Tcp.ConnectState.CONNECTED
import com.wsn.lib.wsb.communicator.tcp.Tcp.ConnectState.CONNECTING
import com.wsn.lib.wsb.communicator.tcp.Tcp.ConnectState.UNCONNECTED

/**
 * Created by CJQ on 2018/3/9.
 */

class TcpServer : Tcp() {

    private var serverSocket: ServerSocket? = null

    val isLaunched: Boolean
        get() = serverSocket != null

    fun launch(localPort: Int): Boolean {
        if (!isLaunched) {
            try {
                serverSocket = ServerSocket(localPort)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        return isLaunched
    }

    fun accept(listener: OnClientAcceptListener) {
        if (isLaunched && state != CONNECTING) {
            state = CONNECTING
            Thread(Runnable {
                while (isLaunched && state == CONNECTING) {
                    var tcpSocket: TcpSocket? = null
                    try {
                        val client = serverSocket!!.accept()
                        tcpSocket = TcpSocket(client)
                    } catch (e: Exception) {
                        e.printStackTrace()
                    } finally {
                        state = if (tcpSocket != null) CONNECTED else UNCONNECTED
                        if (listener.onClientAccept(state, tcpSocket)) {
                            state = CONNECTING
                        }
                    }
                }
            }).start()
        } else {
            listener.onClientAccept(state, null)
        }
    }

    fun shutdown() {
        if (isLaunched) {
            try {
                if (serverSocket?.isClosed == false) {
                    serverSocket?.close()
                }
            } catch (e: IOException) {
                e.printStackTrace()
            } finally {
                state = UNCONNECTED
                serverSocket = null
            }
        }
    }

    /**
     * Created by CJQ on 2018/3/12.
     */
    interface OnClientAcceptListener {
        //返回true将继续等待客户端连接
        fun onClientAccept(state: Tcp.ConnectState, socket: TcpSocket?): Boolean
    }
}
